diffrax/integrate.py

```diff
@@ -142,34 +180,108 @@ def body_fun(state, inplace):
+                min_delay.append(delay(state.tprev, state.y, args))
+            min_delay = jnp.stack(min_delay).min()
+            implicit_step = min_delay < state.tnext - state.tprev
+
+            def history_extrapolation_implicit():
+                implicit_ts = state.dense_ts.at[state.dense_save_index].set(state.tnext)
+                eps = 10e-3
+
+                def _cond_fun(_val):
+                    _, _pred, _ = _val
+                    return _pred
+
+                def _body_fun(_val):
+                    init_dense_infos, _, step = _val
+                    init_dense_interp = DenseInterpolation(
+                        ts=implicit_ts,
+                        ts_size=state.dense_save_index + 1,
+                        interpolation_cls=solver.interpolation_cls,
+                        infos=init_dense_infos,
+                        direction=1,
+                    )
+                    is_vf_wrapper = lambda x: isinstance(x, VectorFieldWrapper)
+
+                    def _apply_history(x):
+                        if is_vf_wrapper(x):
+                            vector_field = _HistoryVectorField(
+                                x.vector_field,
+                                init_dense_interp,
+                                y0_history,
+                                delays,
+                                t0,
+                            )
+                            return VectorFieldWrapper(vector_field)
+                        else:
+                            return x
+
+                    terms_ = jtu.tree_map(_apply_history, terms, is_leaf=is_vf_wrapper)
+                    (y, _, dense_info, _, _) = solver.step(
+                        terms_,
+                        state.tprev,
+                        state.tnext,
+                        state.y,
+                        args,
+                        state.solver_state,
+                        state.made_jump,
+                    )
+
+                    _pred = (
+                        ((y - init_dense_interp.evaluate(state.tnext)) / y) > eps
+                    ).any()
+                    _pred = _pred & (step < 10)
+                    flatt_dense_infos, treedef = jtu.tree_flatten(init_dense_infos)
+                    dense_info, _ = jtu.tree_flatten(dense_info)
+                    new_dense_infos = jtu.tree_map(
+                        lambda x, y: x.at[state.dense_save_index].set(y),
+                        flatt_dense_infos,
+                        dense_info,
+                    )
+                    new_dense_infos = jtu.build_tree(treedef, flatt_dense_infos)
+                    return new_dense_infos, _pred, step + 1
+
+                _init_val = (state.dense_infos, True, 0)
+                val = lax.while_loop(_cond_fun, _body_fun, _init_val)
+                return val[0]
+
+            dense_infos = lax.cond(
+                implicit_step, history_extrapolation_implicit, lambda: state.dense_infos
```

Comment:
What you are saying is interesting, could you develop on that ? Regarding the last time we spoke, what you said (and reminded me in the comment above) is that the `lax.cond` will unconditionally execute both branch regardless of the value of `implicit_step`. 
However, if you add another `lax.cond` inside the the `_cond_fun` of the `lax.while_loop`, the bool value of `implicit_step` won't be branched with JAX (as it did before) ? To this regard, this would mean that `implicit_step` inside the new `lax.cond` would be "concrete" ?

Reply:
So i'll just add a `_pred =lax.cond(implicit_step, lambda : _pred, lambda : False)` inside the `_cond_fun` if that makes sens to you. After testing it seems to shed a coupled of ms most of the time !

Reply:
And that in addition, we should *not* have a `lax.cond` here!

Reply:
makes sense ill swap https://github.com/patrick-kidger/diffrax/pull/169/files#r1011952031 with the latest `lax.cond`